### ISSUE TITLE
Odd behaviour using `map` on `List` with `filter`.

### ISSUE THREAD

**Issue by anlaital** [realm/realm-cocoa#4124]
## Goals

Filter a `List` by mapping from other another `List`.

## Expected Results

Filtering succeeds.

## Actual Results

Filtering crashes with `*** Terminating app due to uncaught exception 'Invalid value', reason: 'IN clause requires an array of items'`.

This is caused by the change to using lazy collections.

If this is the desired outcome, then perhaps having a compile-time error like `NSPredicate` does would be better. 

What is the recommended approach for this scenario with the current Realm version?

## Steps to Reproduce

```
class A: Object {
    
    dynamic var id = ""
    
}

let obs = realm.objects(A.self)
let ids = obs.map { $0.id }
let obs2 = realm.objects(A.self).filter("id IN %@", ids)
```

## Version of Realm and Tooling

Realm version: 1.1.0
Xcode version: 8.0

**Comment by jpsim** [comment]
We've since discussed this on Twitter:https://twitter.com/simjp/status/778620485319942145

In Swift 2.x, `RealmCollectionType` conformed to `Swift.CollectionType`. This was never the right protocol to conform to given Realm's lazy access semantics, since functions on `CollectionType` that perform an operation on the collection was evaluated eagerly for each member of the collection (e.g. `map()`) rather than upon access.

So in Swift 3.0, since the language required us to change our API in a breaking way, we took the opportunity to make `RealmCollection` now conform to `LazyCollectionProtocol` which is a better match for our semantics.

When we made this change, we evaluated if it would break any existing code, and our take on it at the time was that it would not, so we made the change.

Code that relies on the result of `map`/`filter` returning an `Array` (or something that bridges to `NSArray`) will break because of this change, and passing `obs.map { $0.id }` to an `NSPredicate` is one of those cases.

Because we've already shipped this change, I don't think we'll be reverting it, especially since it's possible to get the old behavior on top of the new one, but not the other way around. For example, changing your code sample to this should work, eagerly evaluating the `LazyCollectionProtocol` to an `Array`:

```swift
class A: Object {
  dynamic var id = ""
}

let obs = realm.objects(A.self)
let ids = obs.map { $0.id }
let obs2 = realm.objects(A.self).filter("id IN %@", Array(ids))
```
